crates/noirc_evaluator/src/ssa/acir_gen.rs

```diff
@@ -0,0 +1,720 @@
+use super::node::{Instruction, Operation};
+use acvm::FieldElement;
+use arena::Index;
+use num_traits::ToPrimitive;
+use std::collections::HashMap;
+//use crate::acir::native_types::{Arithmetic, Witness};
+use crate::ssa::code_gen::IRGenerator;
+use crate::ssa::node;
+use crate::ssa::node::Node;
+use crate::Evaluator;
+use crate::Gate;
+use crate::RuntimeErrorKind;
+use acvm::acir::circuit::gate::Directive;
+use acvm::acir::native_types::{Arithmetic, Linear, Witness};
+use num_bigint::BigUint;
+use std::convert::TryInto;
+pub struct Acir {
+    pub arith_cache: HashMap<Index, InternalVar>,
+}
+
+#[derive(Clone, Debug)]
+pub struct InternalVar {
+    expression: Arithmetic,
+    //value: FieldElement,     //not used for now
+    witness: Option<Witness>,
+    idx: Option<Index>,
+}
+impl InternalVar {
+    pub fn is_equal(&self, b: &InternalVar) -> bool {
+        if self.idx.is_some() && b.idx.is_some() && self.idx.unwrap() == b.idx.unwrap() {
+            return true;
+        }
+        if self.witness.is_some()
+            && b.witness.is_some()
+            && self.witness.unwrap() == b.witness.unwrap()
+        {
+            return true;
+        }
+        false //TODO should we check if content is the same??
```

Comment:
I have added the content check, in fact it will work with or without but it may impact on the performance as doing the content check is more expensive.